### DISCUSSION TITLE
new effective porosity and effective permeability model implementation

### DISCUSSION THREAD

**Discussion by rksin8** [idaholab/moose#19962] [Q&A Modules: General] [answered]
Hi,
I want to implement these two relations for effective porosity and effective permeability respectively:
<img src="https://render.githubusercontent.com/render/math?math=\phi_{eff} = \frac{\phi_0 -\epsilon_v}{1-\epsilon_v}">  
 <img src="https://render.githubusercontent.com/render/math?math=\k_{eff} =  k_0 \exp \left( 22.2 (\frac{\phi_{eff}}{\phi_0}-1)\right)">

I modified the existing moose implementation as: 

1) Effective porosity:

```
void
PorosityFromStrainDumux::computeQpProperties()
{
  const unsigned qp_to_use =
      (_nodal_material && _bnd  ? nearestQP(_qp) : _qp);

  _porosity[_qp] = (_initial_porosity - (*_vol_strain_qp)[qp_to_use] )/(1.0 -(*_vol_strain_qp)[qp_to_use]);

  if (_porosity[_qp] < _residual_porosity)
  {
     _porosity[_qp] = _residual_porosity;
  }
  _dporosity_dvar[_qp].resize(_num_var);
  for (unsigned int v = 0; v < _num_var; ++v)
    _dporosity_dvar[_qp][v] = 0;

  _dporosity_dgradvar[_qp].resize(_num_var);
  for (unsigned int v = 0; v < _num_var; ++v)
    _dporosity_dgradvar[_qp][v] = 0;
}
```

2) Effective permeability:
```
void
PermeabilityKozenyCarmanRutqvist::computeQpProperties()
{
  _permeability_qp[_qp] =
      _k_anisotropy * _k0 * std::exp(22.2* (_porosity_qp[_qp]/_phi0 - 1));

  _dpermeability_qp_dvar[_qp].resize(_num_var, RealTensorValue());
  for (unsigned int v = 0; v < _num_var; ++v)
    _dpermeability_qp_dvar[_qp][v] = _dporosity_qp_dvar[_qp][v] * _permeability_qp[_qp] * 22.2/_phi0;

  _dpermeability_qp_dgradvar[_qp].resize(LIBMESH_DIM);
  for (unsigned i = 0; i < LIBMESH_DIM; ++i)
  {
    _dpermeability_qp_dgradvar[_qp][i].resize(_num_var, RealTensorValue());
    for (unsigned int v = 0; v < _num_var; ++v)
      _dpermeability_qp_dgradvar[_qp][i][v] = 
          _dporosity_qp_dgradvar[_qp][v](i) * _permeability_qp[_qp] * 22.2/_phi0;
  }
}
```
But now having convergence issue.  Would you please confirm these implementations are correct? Thank you!

**Comment by WilkAndy** [comment, marked as answer]
Your perm looks OK - hopefully i haven't missed anything there.  The only problem i foresee is the "22.2": since it's reasonably large, the exponential changes will be huge, and MOOSE will probably experience poor convergence.

Your porosity derivatives are incorrect.  They need to include things like (*_dvol_strain_qp_dvar)[qp_to_use][pvar].  You can see examples of this in PorousFlowPorosity and PorousFlowPorosityExponentialBase.  This is probably the cause of your convergence problems.  However, another problem is the cutoff at residual porosity: any "if" conditions like you have are likely to cause convergence problems because of the "sharp corner" at residual porosity.
